ssl/ssl_lib.c

```diff
@@ -2618,8 +2618,11 @@ const char *SSL_get_servername(const SSL *s, const int type)
      * peer send" and "what was actually negotiated"; we should have
      * a clear distinction amongst those three.
      */
-    if (SSL_in_init(s))
+    if (SSL_in_init(s)) {
+        if (s->hit)
+            return s->session->ext.hostname;
```

Comment:
Do you need to check `s->session != NULL` as is done below?

Reply:
I believe that it's impossible to have a resumed SSL with no associated SSL_SESSION.
(But I'm happy to add a check if we don't want to rely on that as an invariant.)

Reply:
True, you have the check on `s->hit`...